### DISCUSSION TITLE
where's the lua API documentation for plugin developer

### DISCUSSION THREAD

**Discussion by caibirdme** [apache/apisix#5421] [Q&A] [answered]
I'm trying to learn to develop apisix plugin in lua, and I see the examples in documentation.
But in that example, it import lua lib, like:
```lua
local core = require("apisix.core")
local router = require("apisix.router")
local ngx = ngx
-- ...
```
But I don't know what APIs these lib provide, and I can't find docs. The only way is to look into the source code?

**Comment by tzssangglass** [comment, marked as answer]
Yes, reading the source code is the best way. See also other plugins code.

**Reply by juzhiyuan** [reply]
At the beginning of developing plugins or debugging codes, It's a little confusing for me as well 😂 We are looking forward to the Best Practices on Developing Apache APISIX, but not much time on this for me, @caibirdme You could have a try and share some experiences on this, if you have any questions during development, kindly let us know :)

**Reply by tzssangglass** [reply]
I think it's time to include a tutorial on plugin development (including documentation and tools) as one of the goals of the community development. Would like to hear the views of others.

**Reply by juzhiyuan** [reply]
Ya, and I remember there has one rule called `don't make me think` 😄

**Reply by caibirdme** [reply]
Thx, and I'm doing so. But I still hope that apisix can provide more detailed docs for plugin development. If I don't have a detailed api docs, I don't know if I could implement some features, thus one will be hesitate about investing time into apisix

**Reply by juzhiyuan** [reply]
Agree with you, and the community needs more contributions from contributors. If you have time, I sincerely invite you and more members to join Apache APISIX's community :)

**Reply by caibirdme** [reply]
When I'm reading about the source code, I'll try 2 make some docs. Is there any api doc format or so, that openresty or apisix is using

**Reply by tzssangglass** [reply]
just like:https://github.com/api7/lua-resty-etcd/blob/master/api_v3.md#methods

It would be good to have an example of how to use it.
